File: projects/animation-engine/tests/test_easing.py
"""Tests for easing functions."""

import math
import pytest
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

from animation_engine.easing import (
    linear,
    ease_in_quad, ease_out_quad, ease_in_out_quad,
    ease_in_cubic, ease_out_cubic, ease_in_out_cubic,
    ease_in_sine, ease_out_sine, ease_in_out_sine,
    ease_in_expo, ease_out_expo, ease_in_out_expo,
    ease_in_circ, ease_out_circ, ease_in_out_circ,
    ease_in_elastic, ease_out_elastic, ease_in_out_elastic,
    ease_in_back, ease_out_back, ease_in_out_back,
    ease_in_bounce, ease_out_bounce, ease_in_out_bounce,
    get_easing_function, list_easing_functions,
    EASING_FUNCTIONS,
)


class TestLinearEasing:
    def test_start(self):
        assert linear(0.0) == 0.0

    def test_end(self):
        assert linear(1.0) == 1.0

    def test_mid(self):
        assert linear(0.5) == 0.5

    def test_quarter(self):
        assert linear(0.25) == 0.25


class TestQuadraticEasing:
    def test_ease_in_quad_boundaries(self):
        assert ease_in_quad(0.0) == 0.0
        assert ease_in_quad(1.0) == 1.0

    def test_ease_in_quad_mid(self):
        assert ease_in_quad(0.5) == pytest.approx(0.25)

    def test_ease_out_quad_boundaries(self):
        assert ease_out_quad(0.0) == 0.0
        assert ease_out_quad(1.0) == pytest.approx(1.0)

    def test_ease_out_quad_mid(self):
        assert ease_out_quad(0.5) == pytest.approx(0.75)

    def test_ease_in_out_quad_boundaries(self):
        assert ease_in_out_quad(0.0) == 0.0
        assert ease_in_out_quad(1.0) == pytest.approx(1.0)

    def test_ease_in_out_quad_mid(self):
        assert ease_in_out_quad(0.5) == pytest.approx(0.5)


class TestCubicEasing:
    def test_ease_in_cubic_boundaries(self):
        assert ease_in_cubic(0.0) == 0.0
        assert ease_in_cubic(1.0) == 1.0

    def test_ease_in_cubic_mid(self):
        assert ease_in_cubic(0.5) == pytest.approx(0.125)

    def test_ease_out_cubic_boundaries(self):
        assert ease_out_cubic(0.0) == pytest.approx(0.0)
        assert ease_out_cubic(1.0) == pytest.approx(1.0)

    def test_ease_in_out_cubic_boundaries(self):
        assert ease_in_out_cubic(0.0) == 0.0
        assert ease_in_out_cubic(1.0) == pytest.approx(1.0)


class TestSineEasing:
    def test_ease_in_sine_boundaries(self):
        assert ease_in_sine(0.0) == 0.0
        assert ease_in_sine(1.0) == pytest.approx(1.0)

    def test_ease_out_sine_boundaries(self):
        assert ease_out_sine(0.0) == 0.0
        assert ease_out_sine(1.0) == pytest.approx(1.0)

    def test_ease_in_out_sine_boundaries(self):
        assert ease_in_out_sine(0.0) == 0.0
        assert ease_in_out_sine(1.0) == pytest.approx(1.0)

    def test_ease_in_out_sine_mid(self):
        assert ease_in_out_sine(0.5) == pytest.approx(0.5)


class TestExponentialEasing:
    def test_ease_in_expo_boundaries(self):
        assert ease_in_expo(0.0) == 0.0
        assert ease_in_expo(1.0) == pytest.approx(1.0)

    def test_ease_out_expo_boundaries(self):
        assert ease_out_expo(0.0) == 0.0
        assert ease_out_expo(1.0) == pytest.approx(1.0)

    def test_ease_in_out_expo_boundaries(self):
        assert ease_in_out_expo(0.0) == 0.0
        assert ease_in_out_expo(1.0) == pytest.approx(1.0)


class TestCircularEasing:
    def test_ease_in_circ_boundaries(self):
        assert ease_in_circ(0.0) == 0.0
        assert ease_in_circ(1.0) == pytest.approx(1.0)

    def test_ease_out_circ_boundaries(self):
        assert ease_out_circ(0.0) == 0.0
        assert ease_out_circ(1.0) == pytest.approx(1.0)

    def test_ease_in_out_circ_boundaries(self):
        assert ease_in_out_circ(0.0) == 0.0
        assert ease_in_out_circ(1.0) == pytest.approx(1.0)


class TestElasticEasing:
    def test_ease_in_elastic_boundaries(self):
        assert ease_in_elastic(0.0) == 0.0
        assert ease_in_elastic(1.0) == pytest.approx(1.0)

    def test_ease_out_elastic_boundaries(self):
        assert ease_out_elastic(0.0) == 0.0
        assert ease_out_elastic(1.0) == pytest.approx(1.0)

    def test_ease_in_out_elastic_boundaries(self):
        assert ease_in_out_elastic(0.0) == 0.0
        assert ease_in_out_elastic(1.0) == pytest.approx(1.0)

    def test_elastic_overshoot(self):
        # Elastic functions should overshoot at certain points
        v = ease_out_elastic(0.2)
        assert v > 1.0  # Overshoots


class TestBackEasing:
    def test_ease_in_back_boundaries(self):
        assert ease_in_back(0.0) == 0.0
        assert ease_in_back(1.0) == pytest.approx(1.0)

    def test_ease_out_back_boundaries(self):
        assert ease_out_back(0.0) == pytest.approx(0.0)
        assert ease_out_back(1.0) == pytest.approx(1.0)

    def test_ease_in_out_back_boundaries(self):
        assert ease_in_out_back(0.0) == 0.0
        assert ease_in_out_back(1.0) == pytest.approx(1.0)

    def test_back_undershoot(self):
        # Back functions should undershoot at start
        v = ease_in_back(0.2)
        assert v < 0.0


class TestBounceEasing:
    def test_ease_in_bounce_boundaries(self):
        assert ease_in_bounce(0.0) == 0.0
        assert ease_in_bounce(1.0) == pytest.approx(1.0)

    def test_ease_out_bounce_boundaries(self):
        assert ease_out_bounce(0.0) == 0.0
        assert ease_out_bounce(1.0) == pytest.approx(1.0)

    def test_ease_in_out_bounce_boundaries(self):
        assert ease_in_out_bounce(0.0) == 0.0
        assert ease_in_out_bounce(1.0) == pytest.approx(1.0)

    def test_bounce_non_negative(self):
        # Bounce values should always be non-negative
        for i in range(101):
            t = i / 100.0
            assert ease_out_bounce(t) >= 0.0


class TestEasingRegistry:
    def test_get_easing_function_valid(self):
        fn = get_easing_function("linear")
        assert fn(0.5) == 0.5

    def test_get_easing_function_css_format(self):
        fn = get_easing_function("ease-in-quad")
        assert fn(0.5) == pytest.approx(0.25)

    def test_get_easing_function_invalid(self):
        with pytest.raises(ValueError, match="Unknown easing function"):
            get_easing_function("nonexistent")

    def test_list_easing_functions(self):
        funcs = list_easing_functions()
        assert "linear" in funcs
        assert "ease_in_quad" in funcs
        assert "ease_out_bounce" in funcs
        assert len(funcs) >= 30

    def test_all_registered_functions_valid(self):
        for name, fn in EASING_FUNCTIONS.items():
            assert fn(0.0) == pytest.approx(0.0, abs=0.01), f"{name} at t=0"
            assert fn(1.0) == pytest.approx(1.0, abs=0.01), f"{name} at t=1"


class TestEasingMonotonicity:
    """Test that ease-in and ease-out functions are monotonic."""

    def test_ease_in_quad_monotone(self):
        prev = ease_in_quad(0.0)
        for i in range(1, 101):
            t = i / 100.0
            v = ease_in_quad(t)
            assert v >= prev
            prev = v

    def test_ease_out_cubic_monotone(self):
        prev = ease_out_cubic(0.0)
        for i in range(1, 101):
            t = i / 100.0
            v = ease_out_cubic(t)
            assert v >= prev
            prev = v


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: projects/timeline-engine/tests/test_easing.ts
/**
 * Tests for easing functions
 */

import {
  linear,
  easeInQuad,
  easeOutQuad,
  easeInOutQuad,
  easeInCubic,
  easeOutCubic,
  easeInOutCubic,
  easeInQuart,
  easeOutQuart,
  easeInOutQuart,
  easeInQuint,
  easeOutQuint,
  easeInOutQuint,
  easeInSine,
  easeOutSine,
  easeInOutSine,
  easeInExpo,
  easeOutExpo,
  easeInOutExpo,
  easeInCirc,
  easeOutCirc,
  easeInOutCirc,
  easeInElastic,
  easeOutElastic,
  easeInOutElastic,
  easeInBack,
  easeOutBack,
  easeInOutBack,
  easeInBounce,
  easeOutBounce,
  easeInOutBounce,
  resolveEasing,
  getEasingNames,
  registerEasing,
} from '../src/easing';

describe('Easing Functions', () => {
  describe('Boundary Conditions', () => {
    it('all easing functions should map 0->0', () => {
      const easings = [
        linear, easeInQuad, easeOutQuad, easeInOutQuad,
        easeInCubic, easeOutCubic, easeInOutCubic,
        easeInQuart, easeOutQuart, easeInOutQuart,
        easeInQuint, easeOutQuint, easeInOutQuint,
        easeInSine, easeOutSine, easeInOutSine,
        easeInExpo, easeOutExpo, easeInOutExpo,
        easeInCirc, easeOutCirc, easeInOutCirc,
        easeInElastic, easeOutElastic, easeInOutElastic,
        easeInBack, easeOutBack, easeInOutBack,
        easeInBounce, easeOutBounce, easeInOutBounce,
      ];

      for (const easing of easings) {
        expect(easing(0)).toBeCloseTo(0, 10);
      }
    });

    it('all easing functions should map 1->1', () => {
      const easings = [
        linear, easeInQuad, easeOutQuad, easeInOutQuad,
        easeInCubic, easeOutCubic, easeInOutCubic,
        easeInQuart, easeOutQuart, easeInOutQuart,
        easeInQuint, easeOutQuint, easeInOutQuint,
        easeInSine, easeOutSine, easeInOutSine,
        easeInExpo, easeOutExpo, easeInOutExpo,
        easeInCirc, easeOutCirc, easeInOutCirc,
        easeInElastic, easeOutElastic, easeInOutElastic,
        easeInBack, easeOutBack, easeInOutBack,
        easeInBounce, easeOutBounce, easeInOutBounce,
      ];

      for (const easing of easings) {
        expect(easing(1)).toBeCloseTo(1, 10);
      }
    });
  });

  describe('Linear', () => {
    it('should return input unchanged', () => {
      expect(linear(0)).toBe(0);
      expect(linear(0.5)).toBe(0.5);
      expect(linear(1)).toBe(1);
    });
  });

  describe('Quadratic', () => {
    it('easeInQuad should be convex', () => {
      expect(easeInQuad(0.5)).toBeCloseTo(0.25, 5);
      expect(easeInQuad(0.5)).toBeLessThan(0.5);
    });

    it('easeOutQuad should be concave', () => {
      expect(easeOutQuad(0.5)).toBeCloseTo(0.75, 5);
      expect(easeOutQuad(0.5)).toBeGreaterThan(0.5);
    });

    it('easeInOutQuad should be symmetric', () => {
      expect(easeInOutQuad(0.5)).toBeCloseTo(0.5, 5);
    });
  });

  describe('Cubic', () => {
    it('easeInCubic should be convex', () => {
      expect(easeInCubic(0.5)).toBeCloseTo(0.125, 5);
      expect(easeInCubic(0.5)).toBeLessThan(0.5);
    });

    it('easeOutCubic should be concave', () => {
      expect(easeOutCubic(0.5)).toBeCloseTo(0.875, 5);
      expect(easeOutCubic(0.5)).toBeGreaterThan(0.5);
    });

    it('easeInOutCubic should be symmetric', () => {
      expect(easeInOutCubic(0.5)).toBeCloseTo(0.5, 5);
    });
  });

  describe('Sine', () => {
    it('easeInSine should be convex', () => {
      const val = easeInSine(0.5);
      expect(val).toBeLessThan(0.5);
    });

    it('easeOutSine should be concave', () => {
      const val = easeOutSine(0.5);
      expect(val).toBeGreaterThan(0.5);
    });

    it('easeInOutSine should be symmetric', () => {
      expect(easeInOutSine(0.5)).toBeCloseTo(0.5, 5);
    });
  });

  describe('Exponential', () => {
    it('easeInExpo should be very convex', () => {
      const val = easeInExpo(0.5);
      expect(val).toBeLessThan(0.1);
    });

    it('easeOutExpo should be very concave', () => {
      const val = easeOutExpo(0.5);
      expect(val).toBeGreaterThan(0.9);
    });

    it('easeInOutExpo should be symmetric', () => {
      expect(easeInOutExpo(0.5)).toBeCloseTo(0.5, 5);
    });
  });

  describe('Back', () => {
    it('easeInBack should not overshoot below 0 for t in [0,1]', () => {
      const val = easeInBack(0.5);
      expect(val).toBeGreaterThanOrEqual(0);
    });

    it('easeOutBack should have overshoot near end', () => {
      const val = easeOutBack(0.9);
      expect(val).toBeGreaterThan(1);
    });
  });

  describe('Bounce', () => {
    it('easeOutBounce should bounce near 1', () => {
      const val = easeOutBounce(0.9);
      expect(val).toBeGreaterThan(0.9);
      expect(val).toBeLessThan(1.0);
    });

    it('easeInBounce should start slowly', () => {
      const val = easeInBounce(0.1);
      expect(val).toBeLessThan(0.1);
    });

    it('easeInOutBounce should be symmetric', () => {
      expect(easeInOutBounce(0.5)).toBeCloseTo(0.5, 5);
    });
  });

  describe('Elastic', () => {
    it('easeOutElastic should oscillate near 1', () => {
      const val = easeOutElastic(0.9);
      expect(val).toBeLessThan(1.0);
    });

    it('easeOutElastic should oscillate near 1', () => {
      const val = easeOutElastic(0.9);
      expect(val).toBeLessThan(1.0);
    });

    it('easeInOutElastic should be symmetric', () => {
      expect(easeInOutElastic(0.5)).toBeCloseTo(0.5, 5);
    });
  });

  describe('resolveEasing', () => {
    it('should return default for undefined', () => {
      const result = resolveEasing(undefined);
      expect(result).toBeDefined();
      expect(result(0)).toBeCloseTo(0, 10);
      expect(result(1)).toBeCloseTo(1, 10);
    });

    it('should return function for function input', () => {
      const custom = (t: number) => t * t;
      const result = resolveEasing(custom);
      expect(result).toBe(custom);
    });

    it('should resolve named easing', () => {
      const result = resolveEasing('linear');
      expect(result(0.5)).toBeCloseTo(0.5, 10);
    });

    it('should fall back for unknown name', () => {
      const result = resolveEasing('nonexistent-easing');
      expect(result).toBeDefined();
      expect(result(0)).toBeCloseTo(0, 10);
      expect(result(1)).toBeCloseTo(1, 10);
    });
  });

  describe('getEasingNames', () => {
    it('should return non-empty array', () => {
      const names = getEasingNames();
      expect(names.length).toBeGreaterThan(0);
    });

    it('should include common easings', () => {
      const names = getEasingNames();
      expect(names).toContain('linear');
      expect(names).toContain('ease-in-cubic');
      expect(names).toContain('ease-out-cubic');
      expect(names).toContain('ease-in-out-cubic');
    });
  });

  describe('registerEasing', () => {
    it('should register custom easing', () => {
      const custom = (t: number) => t * t;
      registerEasing('custom-quad', custom);
      const names = getEasingNames();
      expect(names).toContain('custom-quad');

      const resolved = resolveEasing('custom-quad');
      expect(resolved(0.5)).toBeCloseTo(0.25, 10);
      expect(resolved(0)).toBeCloseTo(0, 10);
      expect(resolved(1)).toBeCloseTo(1, 10);
    });
  });
});
